pallets/reward-distribution/src/lib.rs

```diff
@@ -94,10 +169,229 @@ pub mod pallet {
 			ensure_root(origin)?;
 
 			RewardPerBlock::<T>::put(new_reward_per_block);
-			RewardsAdaptedAt::<T>::put(frame_system::Pallet::<T>::block_number());
-
+			RewardsAdaptedAt::<T>::put(ext::security::get_active_block::<T>());
 			Self::deposit_event(Event::<T>::RewardPerBlockAdapted(new_reward_per_block));
 			Ok(())
 		}
+
+		/// Allow users who have staked to collect rewards for a given vault and rewraded currency
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as Config>::WeightInfo::collect_reward())]
+		#[transactional]
+		pub fn collect_reward(
+			origin: OriginFor<T>,
+			vault_id: DefaultVaultId<T>,
+			reward_currency_id: CurrencyId<T>,
+			index: Option<T::Index>,
+		) -> DispatchResult {
+			//distribute reward from reward pool to staking pallet store
+			let caller = ensure_signed(origin)?;
+			let reward = ext::pooled_rewards::withdraw_reward::<T>(
+				&vault_id.collateral_currency(),
+				&vault_id,
+				reward_currency_id,
+			)?;
+			ext::staking::distribute_reward::<T>(&vault_id, reward, reward_currency_id)?;
+
+			//withdraw the reward for specific nominator
+			let caller_rewards =
+				ext::staking::withdraw_reward::<T>(&vault_id, &caller, index, reward_currency_id)?;
+
+			if caller_rewards == (BalanceOf::<T>::zero()) {
+				return Err(Error::<T>::NoRewardsForAccount.into())
+			}
+
+			//transfer rewards
+			Self::transfer_reward(reward_currency_id, caller_rewards, caller)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn transfer_reward(
+		reward_currency_id: CurrencyId<T>,
+		reward: BalanceOf<T>,
+		beneficiary: T::AccountId,
+	) -> DispatchResult {
+		let native_currency_id = T::GetNativeCurrencyId::get();
+		if reward_currency_id == native_currency_id {
+			let available_native_funds = T::Balances::free_balance(&Self::fee_pool_account_id());
+
+			match reward.checked_sub(&available_native_funds) {
+				None => {
+					// Pay the whole reward from the fee pool
+					let amount: currency::Amount<T> = Amount::new(reward, reward_currency_id);
+					return amount.transfer(&Self::fee_pool_account_id(), &beneficiary)
+				},
+				Some(remaining) => {
+					//check if the to-be-minted amount is consistent
+					//with reward tracking
+					let liability = NativeLiability::<T>::get()
+						.ok_or(Error::<T>::NotEnoughRewardsRegistered)?;
+
+					if liability < remaining {
+						return Err(Error::<T>::NotEnoughRewardsRegistered.into())
+					}
+
+					// Use the available funds from the fee pool
+					let available_amount: currency::Amount<T> =
+						Amount::new(available_native_funds, reward_currency_id);
+					available_amount.transfer(&Self::fee_pool_account_id(), &beneficiary)?;
+					// Mint the rest
+					T::Balances::deposit_creating(&beneficiary, remaining);
+
```

Comment:
There is a potential bug where the reward is below the existential deposit which I have encountered in testing. 
I say potential given that it depends on how one defines the rewards per block. If they are lower than the existential deposit, a user could attempt to collect native bellow the existential value when it's account does not exists, which will result in burning those rewards essentially.

Since this is a very edge case and the logic is already quite big, do you think we should add a check for this @ebma ?

Reply:
Hmm okay, nice catch.

> If they are lower than the existential deposit, a user could attempt to collect native bellow the existential value when it's account does not exists, which will result in burning those rewards essentially.

But I think a user who does not have enough native tokens to fulfill the existential deposit also would not have enough native tokens to issue the transaction in the first place, or would he?

Reply:
This would be interesting for non-native tokens as well though. IIRC orml-tokens also defines an existential deposit in its config. What would happen if a nominator tries to withdraw some very small amount that is below the exitential deposit of orml-tokens? Something like 100 units should not work probably.

Reply:
The check is done before the withdraw, [here](https://github.com/pendulum-chain/spacewalk/pull/415/files#diff-2f743609293dfd3b50158de75d67dabc3018b7fa2abe78bed8fd45a783aa3975R202)

Reply:
Nice, great stuff 👍